### ISSUE TITLE
Issue with food from City states

### ISSUE THREAD

**Issue by RemitoL** [LoneGazebo/Community-Patch-DLL#10944]
_1. Mod Version (X.Y.Z). Current Version: 4.7.3_
Version 4.7.3

_3. List of Other Mods_
Community Events for VP

_4. Describe the Issue_
I noticed in my latest game that many of my cities were suddenly starving. I was a bit surprised but then I realized it never stopped, even after the population loss.
I looked at the Food details to understand, and it seems the Food from maritime city states is calculated in a wrong way, with very negative numbers.
It appears in the Modern era, worsens in the atomic era nad becomes unplayable in the information era (-107 food per turn !)

______________________________________________________________________________________
_5. Save Game From 1 Turn Before (ALWAYS ATTACH THIS IF POSSIBLE)_
This is not from 1 turn before as the bug is there all along the game but I attach 1 save from the modern era, 1 from the atomic and 1 from the information era
[Pacal_saves.zip](https://github.com/user-attachments/files/15519084/Pacal_saves.zip)

_6. Logs (ALWAYS ATTACH THESE IF POSSIBLE)_
Not sure I have activated those

_9. Screenshots of the Issue (Optional)_
Screenshots from a city (non capital) in modern, atomic and information era

![City screen modern era](https://github.com/LoneGazebo/Community-Patch-DLL/assets/171387951/33bbe082-22b4-4611-81cd-d312abf50a37)
![City screen atomic era](https://github.com/LoneGazebo/Community-Patch-DLL/assets/171387951/0f199144-9112-4c04-bdb7-748a41edddbf)

![City screen information era](https://github.com/LoneGazebo/Community-Patch-DLL/assets/171387951/ec7c71f7-a339-470a-abfb-679c82ad466a)

**Comment by azum4roll** [comment]
Maritime CS yields add to the city tile and shouldn't be represented as "from City-States". Something has to be wrong there.

**Comment by bsesar** [comment]
I don't see negative numbers in my cities, but I can confirm the "from City-States" part. Also, while the food from City States in my other cities matches the tooltip associated with my only Maritime CS ally, Quebec City (3 Food in other cities), the amount of food in my capital, (Rome) from City States should be 8, but the city tooltip says 12. I do have 4 City States as allies, so maybe that's where the additional 4 Food in Capital come from? I do not have any Statecraft policies, just Authority and Fealty trees filled up.

<img width="157" alt="Rome_food" src="https://github.com/LoneGazebo/Community-Patch-DLL/assets/620131/d495c59a-a9ba-4a7d-9bd5-f9fcacabf0b0">
<img width="282" alt="Rome_Maritime_CS_food" src="https://github.com/LoneGazebo/Community-Patch-DLL/assets/620131/6010818a-ca09-4ce5-810d-a5997224ecc0">

I am also playing with 4.7.3 and the only other mod I have is Improved City View (v20). Logs and a savefile are attached.

[Logs.zip](https://github.com/user-attachments/files/15525750/Logs.zip)
[CS_food.zip](https://github.com/user-attachments/files/15525753/CS_food.zip)

**Comment by bsesar** [comment]
@azum4roll, when fixing issue #10667, ilteroi noted that

"bonus food from maritime city states is now booked correctly as food from city states and not just added to the city plot"

https://github.com/LoneGazebo/Community-Patch-DLL/issues/10667#issuecomment-1957808507

@ilteroi, any idea what may be causing discrepant food numbers from CS as in the above examples?

**Comment by ilteroi** [comment]
i would like to see this fix, i suspect it's a missing *100 somewhere?

**Comment by axatin** [comment]
> i would like to see this fix, i suspect it's a missing *100 somewhere?

no, something completely different. edge case of having a sphere of influence while not being allied. every turn, the game thought the ally status had just been removed and subtracted the food bonus

**Comment by RemitoL** [comment]
Thanks !
Indeed I can confirm I had at least 2 Maritime CS under Sphere of Influence that were not allied at the time of the vote.

**Comment by bsesar** [comment]
Nice to see that fixed 🙂 Is the 12 Food in Rome (see screenshot above) expected behavior given that I have only one CS ally that gives 8 food?

**Comment by axatin** [comment]
@bsesar Do you have Scrivener‘s Office in Rome?

**Comment by bsesar** [comment]
I do. That explains it, thanks! 🙂

**Comment by azum4roll** [comment]
> > i would like to see this fix, i suspect it's a missing *100 somewhere?
> 
> no, something completely different. edge case of having a sphere of influence while not being allied. every turn, the game thought the ally status had just been removed and subtracted the food bonus

Didn't something similar happen before, but only subtracted once on sphere gain?

**Comment by ilteroi** [comment]
i'm not sure about this fix, did you test it?

look at the checks a few lines above:

```
	// Resolve Allies status with sphere of influence or open door
	bool bHasOtherPermanentAlly = false;
	if (IsNoAlly() || (GetPermanentAlly() != NO_PLAYER && GetPermanentAlly() != ePlayer))
	{
		bNowAllies = false;
		bHasOtherPermanentAlly = true;
	}
```

the ```GetPermanentAlly() != ePlayer``` case is already covered?

**Comment by axatin** [comment]
these two are not in the same if-else block.

**Comment by ilteroi** [comment]
1. bHasOtherPermanentAlly is the link
2. you did not answer the question

**Comment by axatin** [comment]
That change works fine, I tested it. But I agree it's not the most elegant solution, I'll rewrite it.

Looking into the topic a bit deeper, there are also some other bugs when open doors / SoI are set or removed. Working on a fix.

**Comment by ilteroi** [comment]
please wait a while i have also rewritten a lot of this setAlly logic, will push later today

**Comment by axatin** [comment]
ok, let me know when it's pushed
